File: compat/suites/go-sdk/internal/groups/s3tables_schemav2.go
package groups

import (
	"context"
	"fmt"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3tables"
	"github.com/aws/aws-sdk-go-v2/service/s3tables/document"
	"github.com/aws/aws-sdk-go-v2/service/s3tables/types"

	"github.com/overcast-sh/overcast-compat-go-sdk/internal/clients"
	"github.com/overcast-sh/overcast-compat-go-sdk/internal/harness"
)

// S3TablesSchemaV2 returns the s3tables-schemav2 group: CreateTable with a
// nested schemaV2 — a list, a map and a struct as Iceberg type documents —
// partitioned by a field of the struct, then the metadata file it wrote.
func S3TablesSchemaV2(c *clients.Clients) ServiceGroup {
	g := &s3tablesSchemaV2Group{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"s3tables-schemav2:CreateTable":              g.CreateTable,
			"s3tables-schemav2:GetTableMetadataLocation": g.GetTableMetadataLocation,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"s3tables-schemav2": g.setup,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"s3tables-schemav2": g.teardown,
		},
	}
}

type s3tablesSchemaV2Group struct{ c *clients.Clients }

const s3tablesSchemaV2Table = "events"

func (g *s3tablesSchemaV2Group) cl() *s3tables.Client { return g.c.S3Tables() }

func (g *s3tablesSchemaV2Group) bucket(t *harness.TestContext) string {
	return tableBucketName("s3tables-v2-", t)
}

// namespace allows underscores, not hyphens.
func (g *s3tablesSchemaV2Group) namespace(t *harness.TestContext) string {
	return strings.ReplaceAll(g.bucket(t), "-", "_")
}

func (g *s3tablesSchemaV2Group) table(t *harness.TestContext) (*string, *string, *string) {
	return aws.String(t.GetString("s3tablesV2BucketARN")), aws.String(g.namespace(t)), aws.String(s3tablesSchemaV2Table)
}

func (g *s3tablesSchemaV2Group) setup(ctx context.Context, t *harness.TestContext) error {
	b, err := g.cl().CreateTableBucket(ctx, &s3tables.CreateTableBucketInput{Name: aws.String(g.bucket(t))})
	if err != nil {
		return err
	}
	t.Set("s3tablesV2BucketARN", aws.ToString(b.Arn))
	_, err = g.cl().CreateNamespace(ctx, &s3tables.CreateNamespaceInput{TableBucketARN: b.Arn, Namespace: []string{g.namespace(t)}})
	return err
}

func (g *s3tablesSchemaV2Group) teardown(ctx context.Context, t *harness.TestContext) error {
	arn, ns, name := g.table(t)
	if aws.ToString(arn) == "" {
		return nil
	}
	g.cl().DeleteTable(ctx, &s3tables.DeleteTableInput{TableBucketARN: arn, Namespace: ns, Name: name}) //nolint:errcheck
	g.cl().DeleteNamespace(ctx, &s3tables.DeleteNamespaceInput{TableBucketARN: arn, Namespace: ns})     //nolint:errcheck
	g.cl().DeleteTableBucket(ctx, &s3tables.DeleteTableBucketInput{TableBucketARN: arn})                //nolint:errcheck
	return nil
}

func (g *s3tablesSchemaV2Group) CreateTable(ctx context.Context, t *harness.TestContext) error {
	arn, ns, name := g.table(t)
	out, err := g.cl().CreateTable(ctx, &s3tables.CreateTableInput{
		TableBucketARN: arn, Namespace: ns, Name: name, Format: types.OpenTableFormatIceberg,
		Metadata: &types.TableMetadataMemberIceberg{Value: types.IcebergMetadata{
			SchemaV2: &types.IcebergSchemaV2{
				Type: types.SchemaV2FieldTypeStruct, IdentifierFieldIds: []int32{1},
				Fields: []types.SchemaV2Field{
					{Id: aws.Int32(1), Name: aws.String("id"), Required: aws.Bool(true), Type: document.NewLazyDocument("long")},
					{Id: aws.Int32(2), Name: aws.String("tags"), Required: aws.Bool(false), Type: document.NewLazyDocument(map[string]any{
						"type": "list", "element-id": 5, "element": "string", "element-required": false,
					})},
					{Id: aws.Int32(3), Name: aws.String("attributes"), Required: aws.Bool(false), Type: document.NewLazyDocument(map[string]any{
						"type": "map", "key-id": 6, "key": "string", "value-id": 7, "value": "string", "value-required": false,
					})},
					{Id: aws.Int32(4), Name: aws.String("customer"), Required: aws.Bool(true), Type: document.NewLazyDocument(map[string]any{
						"type": "struct", "fields": []any{
							map[string]any{"id": 8, "name": "region", "required": true, "type": "string"},
						},
					})},
				},
			},
			PartitionSpec: &types.IcebergPartitionSpec{Fields: []types.IcebergPartitionField{
				{SourceId: aws.Int32(8), Transform: aws.String("identity"), Name: aws.String("region")},
			}},
		}},
	})
	if err != nil {
		return err
	}
	if !strings.Contains(aws.ToString(out.TableARN), "/table/") || aws.ToString(out.VersionToken) == "" {
		return fmt.Errorf("CreateTable: tableARN %q, versionToken %q", aws.ToString(out.TableARN), aws.ToString(out.VersionToken))
	}
	return nil
}

func (g *s3tablesSchemaV2Group) GetTableMetadataLocation(ctx context.Context, t *harness.TestContext) error {
	arn, ns, name := g.table(t)
	out, err := g.cl().GetTableMetadataLocation(ctx, &s3tables.GetTableMetadataLocationInput{TableBucketARN: arn, Namespace: ns, Name: name})
	if err != nil {
		return err
	}
	loc, warehouse := aws.ToString(out.MetadataLocation), aws.ToString(out.WarehouseLocation)
	if warehouse == "" || !strings.HasPrefix(loc, warehouse+"/metadata/") || !strings.HasSuffix(loc, ".metadata.json") {
		return fmt.Errorf("GetTableMetadataLocation: metadataLocation %q, warehouseLocation %q", loc, warehouse)
	}
	return nil
}

File: compat/suites/python-sdk/groups/s3tables_schemav2.py
"""
groups/s3tables_schemav2.py — S3 Tables CreateTable with a nested schemaV2, for the Python suite.

s3tables-schemav2 creates a table whose schemaV2 holds a list, a map and a
struct as Iceberg type documents, partitioned by a field of the struct, then
reads back the metadata file CreateTable wrote.
"""

from __future__ import annotations

from lib.harness import TestContext
from lib.clients import make_clients

from groups.glue_s3tables import table_bucket_name

_TABLE = "events"

_SCHEMA_V2 = {
    "type": "struct",
    "identifierFieldIds": [1],
    "fields": [
        {"id": 1, "name": "id", "required": True, "type": "long"},
        {"id": 2, "name": "tags", "required": False,
         "type": {"type": "list", "element-id": 5, "element": "string", "element-required": False}},
        {"id": 3, "name": "attributes", "required": False,
         "type": {"type": "map", "key-id": 6, "key": "string", "value-id": 7, "value": "string", "value-required": False}},
        {"id": 4, "name": "customer", "required": True,
         "type": {"type": "struct", "fields": [{"id": 8, "name": "region", "required": True, "type": "string"}]}},
    ],
}


def _s3tables(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region)._get("s3tables")


def _bucket(ctx: TestContext) -> str:
    return table_bucket_name("s3tables-v2-", ctx.run_id)


def _table(ctx: TestContext) -> dict:
    """The group's table, as the S3 Tables calls address it; namespaces allow underscores, not hyphens."""
    return {"tableBucketARN": ctx.get("s3tables_v2_bucket_arn"), "namespace": _bucket(ctx).replace("-", "_"), "name": _TABLE}


def _setup(ctx: TestContext) -> None:
    arn = _s3tables(ctx).create_table_bucket(name=_bucket(ctx))["arn"]
    ctx["s3tables_v2_bucket_arn"] = arn
    _s3tables(ctx).create_namespace(tableBucketARN=arn, namespace=[_table(ctx)["namespace"]])


def _teardown(ctx: TestContext) -> None:
    if not ctx.get("s3tables_v2_bucket_arn"):
        return
    table, client = _table(ctx), _s3tables(ctx)
    for call in (
        lambda: client.delete_table(**table),
        lambda: client.delete_namespace(tableBucketARN=table["tableBucketARN"], namespace=table["namespace"]),
        lambda: client.delete_table_bucket(tableBucketARN=table["tableBucketARN"]),
    ):
        try:
            call()
        except Exception:
            pass


# ── s3tables-schemav2 ─────────────────────────────────────────────────────────

def CreateTable(ctx: TestContext) -> None:
    resp = _s3tables(ctx).create_table(
        **_table(ctx), format="ICEBERG",
        metadata={"iceberg": {
            "schemaV2": _SCHEMA_V2,
            "partitionSpec": {"fields": [{"sourceId": 8, "transform": "identity", "name": "region"}]},
        }},
    )
    if "/table/" not in resp.get("tableARN", "") or not resp.get("versionToken"):
        raise AssertionError(f"CreateTable: tableARN {resp.get('tableARN')!r}, versionToken {resp.get('versionToken')!r}")


def GetTableMetadataLocation(ctx: TestContext) -> None:
    resp = _s3tables(ctx).get_table_metadata_location(**_table(ctx))
    loc, warehouse = resp.get("metadataLocation", ""), resp.get("warehouseLocation", "")
    if not warehouse or not loc.startswith(warehouse + "/metadata/") or not loc.endswith(".metadata.json"):
        raise AssertionError(f"GetTableMetadataLocation: metadataLocation {loc!r}, warehouseLocation {warehouse!r}")


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "s3tables-schemav2:CreateTable": CreateTable,
    "s3tables-schemav2:GetTableMetadataLocation": GetTableMetadataLocation,
}

SETUP = {
    "s3tables-schemav2": _setup,
}
TEARDOWN = {
    "s3tables-schemav2": _teardown,
}
